pkg/kv/kvserver/mvcc_gc_queue.go

```diff
@@ -722,9 +782,78 @@ func updateStoreMetricsWithGCInfo(metrics *StoreMetrics, info gc.Info) {
 	metrics.GCFailedClearRange.Inc(int64(info.ClearRangeKeyFailures))
 }
 
+func (mgcq *mvccGCQueue) postProcessScheduled(
+	ctx context.Context, processedReplica replicaInQueue, priority float64,
+) {
+	if priority < deleteRangePriority {
+		mgcq.lastRangeWasHighPriority = false
+		return
+	}
+
+	if !mgcq.lastRangeWasHighPriority {
+		// We are most likely processing first range that has a GC hint notifying
+		// that multiple range deletions happen.
+		mgcq.scanReplicasForHiPriGCHints(ctx, processedReplica.GetRangeID())
+		// Set flag indicating that we are already collecting high priority to avoid
+		// rescanning and re-enqueueing ranges multiple times.
+		mgcq.lastRangeWasHighPriority = true
+	}
+}
+
+func (mgcq *mvccGCQueue) scanReplicasForHiPriGCHints(
+	ctx context.Context, triggerRange roachpb.RangeID,
+) {
+	// We run scanner for a limited time to ensure GC queue can make progress if
+	// something goes wrong with leaseholders. Since visitor is shuffling replicas
+	// prior to iteration we should eventually process all even if first attempt
+	// fails.
+	var foundRanges int
+	err := contextutil.RunWithTimeout(ctx, "gc-check-hinted-hi-pri-replicas", gcHintScannerTimeout, func(ctx context.Context) error {
```

Comment:
I see a problem with lifting it to baseQueue.processQueue as it'll run and unconditional empty task which is optional for all queues but GC in narrow set of cases.
In that case it would make more sense to have this callback as a function on a queue rather than queue impl interface, but that is ugly as you have to use two separate mechanisms for implementing queue functionality. I think noting that it shouldn't run long would be enough and let queue manage the timeout by itself.
I lifted it to postProcessScheduled.

Reply:
I'm not super-worried about running an empty task, since it's just a function call. But sure, lifting it to the GC queue is fine too.